File: src/main/java/itmo/is/project/model/resource/ResourceAmount.java
package itmo.is.project.model.resource;

import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotNull;
import lombok.*;

@ToString
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode
public class ResourceAmount implements ResourceAmountHolder {

    @NotNull
    private Resource resource;

    @With
    @NotNull
    @Min(1)
    private Integer amount;

    public ResourceAmount(Number resourceId, String resourceName, Number amount) {
        this.resource = new Resource(resourceId.intValue(), resourceName);
        this.amount = amount.intValue();
    }

    public ResourceIdAmount getResourceIdAmount() {
        return new ResourceIdAmount(resource.getId(), amount);
    }

    @Override
    public ResourceAmount getResourceAmount() {
        return this;
    }
}

File: src/main/kotlin/itmo/isproject/shared/resource/model/ResourceAmount.kt
package itmo.isproject.shared.resource.model

interface ResourceIdAmountHolder {
    val resourceId: Int?
    var amount: Int?

    operator fun plusAssign(amount: Int) {
        this.amount = this.amount?.plus(amount) ?: amount
    }

    operator fun minusAssign(amount: Int) {
        this.amount = this.amount?.minus(amount) ?: -amount
    }
}

data class ResourceIdAmount(
    override val resourceId: Int? = null,
    override var amount: Int? = null
) : ResourceIdAmountHolder

interface ResourceAmountHolder : ResourceIdAmountHolder {
    var resource: Resource?
    override val resourceId: Int?
        get() = resource?.id
}

data class ResourceAmount(
    override var resource: Resource? = null,
    override var amount: Int? = null
) : ResourceAmountHolder {

    constructor(resourceId: Number, resourceName: String, amount: Number) : this(
        Resource().apply {
            id = resourceId.toInt()
            name = resourceName
        },
        amount.toInt()
    )
}
